src/rp2_common/hardware_uart/uart.c

```diff
@@ -86,19 +122,80 @@ uint uart_set_baudrate(uart_inst_t *uart, uint baudrate) {
         baud_fbrd = ((baud_rate_div & 0x7f) + 1) / 2;
     }
 
-    // Load PL011's baud divisor registers
+    // Need to cleanly disable UART before touching LCR
+    bool was_enabled = uart_is_enabled(uart);
+    uint32_t cr_save;
+    if (was_enabled)
+        cr_save = uart_disable_before_lcr_write(uart);
+
     uart_get_hw(uart)->ibrd = baud_ibrd;
     uart_get_hw(uart)->fbrd = baud_fbrd;
-
-    // PL011 needs a (dummy) line control register write to latch in the
-    // divisors. We don't want to actually change LCR contents here.
+    // PL011 needs a (dummy) LCR_H write to latch in the divisors. We don't
+    // want to actually change LCR_H contents here.
     hw_set_bits(&uart_get_hw(uart)->lcr_h, 0);
 
+    // Re-enable using saved control register value
+    if (was_enabled)
+        uart_get_hw(uart)->cr = cr_save;
+
     // See datasheet
     return (4 * clock_get_hz(clk_peri)) / (64 * baud_ibrd + baud_fbrd);
 }
 /// \end::uart_set_baudrate[]
 
+void uart_set_format(uart_inst_t *uart, uint data_bits, uint stop_bits, uart_parity_t parity) {
+    invalid_params_if(UART, data_bits < 5 || data_bits > 8);
+    invalid_params_if(UART, stop_bits != 1 && stop_bits != 2);
+    invalid_params_if(UART, parity != UART_PARITY_NONE && parity != UART_PARITY_EVEN && parity != UART_PARITY_ODD);
+
+    bool was_enabled = uart_is_enabled(uart);
+    uint32_t cr_save;
+    if (was_enabled)
+        cr_save = uart_disable_before_lcr_write(uart);
+
+    hw_write_masked(&uart_get_hw(uart)->lcr_h,
+                   ((data_bits - 5u) << UART_UARTLCR_H_WLEN_LSB) |
+                   ((stop_bits - 1u) << UART_UARTLCR_H_STP2_LSB) |
+                   (bool_to_bit(parity != UART_PARITY_NONE) << UART_UARTLCR_H_PEN_LSB) |
+                   (bool_to_bit(parity == UART_PARITY_EVEN) << UART_UARTLCR_H_EPS_LSB),
+                   UART_UARTLCR_H_WLEN_BITS |
+                   UART_UARTLCR_H_STP2_BITS |
+                   UART_UARTLCR_H_PEN_BITS |
+                   UART_UARTLCR_H_EPS_BITS);
+
+    if (was_enabled)
+        uart_get_hw(uart)->cr = cr_save;
+}
```

Comment:
I wonder if it's worth abstracting this out further into e.g. a `uart_write_lcr_bits(uart_inst_t *uart, uint32_t values, uint32_t write_mask)` function, so that `uart_set_format` would just be:
```c
void uart_set_format(uart_inst_t *uart, uint data_bits, uint stop_bits, uart_parity_t parity) {
    invalid_params_if(UART, data_bits < 5 || data_bits > 8);
    invalid_params_if(UART, stop_bits != 1 && stop_bits != 2);
    invalid_params_if(UART, parity != UART_PARITY_NONE && parity != UART_PARITY_EVEN && parity != UART_PARITY_ODD);

    uart_write_lcr_bits(uart,
                       ((data_bits - 5u) << UART_UARTLCR_H_WLEN_LSB) |
                       ((stop_bits - 1u) << UART_UARTLCR_H_STP2_LSB) |
                       (bool_to_bit(parity != UART_PARITY_NONE) << UART_UARTLCR_H_PEN_LSB) |
                       (bool_to_bit(parity == UART_PARITY_EVEN) << UART_UARTLCR_H_EPS_LSB),
                       UART_UARTLCR_H_WLEN_BITS |
                       UART_UARTLCR_H_STP2_BITS |
                       UART_UARTLCR_H_PEN_BITS |
                       UART_UARTLCR_H_EPS_BITS);
}
```
and `uart_set_fifo_enabled` would just be:
```c
void uart_set_fifo_enabled(uart_inst_t *uart, bool enabled) {
    uart_write_lcr_bits(uart,
                       (bool_to_bit(enabled) << UART_UARTLCR_H_FEN_LSB),
                       UART_UARTLCR_H_FEN_BITS);
}
```
:man_shrugging: 

(or perhaps this is one of those cases where too much abstraction is confusing rather than helpful)